### PR TITLE
JDK-8296784: Provide clean mallinfo/mallinfo2 wrapper for Linux glibc platforms

### PR DESCRIPTION
**Pull request by tstuefe**
This is a breakout from [JDK-8293114](https://bugs.openjdk.org/browse/JDK-8293114), which is starved for reviews. So I attempt to break up that fix into smaller units which are hopefully easier to review separately.

We use the glibc-specific mallinfo/mallinfo2 APIs to query libc internals. These are quite important, since they are the only way to get the malloc load and retained size from the glibc allocator. NMT only reports a tiny part of that (only the user-view of whatever hotspot allocated).

Using these APIs and staying glibc-agnostic requires some twiddling. This patch provides a clean wrapper to hide that twiddling from the end caller.

Patch also fixes a display error that may have caused omission of the "may have wrapped" marker in the hs-err printout of the malloc load.

Also, gtest regression test.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296784](https://bugs.openjdk.org/browse/JDK-8296784): Provide clean mallinfo/mallinfo2 wrapper for Linux glibc platforms

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [6dd6ec9b](https://git.openjdk.org/jdk/pull/11088/files/6dd6ec9b8c452e3b1c8b5b9d4d78fca8851ae1d5)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [6dd6ec9b](https://git.openjdk.org/jdk/pull/11088/files/6dd6ec9b8c452e3b1c8b5b9d4d78fca8851ae1d5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11088/head:pull/11088` \
`$ git checkout pull/11088`

Update a local copy of the PR: \
`$ git checkout pull/11088` \
`$ git pull https://git.openjdk.org/jdk pull/11088/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11088`

View PR using the GUI difftool: \
`$ git pr show -t 11088`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11088.diff">https://git.openjdk.org/jdk/pull/11088.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
x86 test error in compiler/c2/TestVerifyGraphEdges.java unrelated

File: src/hotspot/os/linux/os_linux.hpp [hunk 1]
```diff
@@ -426,6 +392,39 @@ class os::Linux {
...
+  // to keep the JVM runtime-compatible with different glibc versions.
+  //
+  // mallinfo2() was added with glibc (>2.32). Legacy mallinfo() was deprecated with
+  // 2.33 and may vanish in future glibcs. So we may have both or either one of
+  // them.
+  //
+  // mallinfo2() is functionally equivalent to legacy mallinfo but returns sizes as
+  // 64-bit on 64-bit platforms. Legacy mallinfo uses 32-bit fields. However, legacy
+  // mallinfo is still perfectly fine to use if we know the sizes cannot have wrapped.
+  // For example, if the process virtual size does not exceed 4G, we cannot hav
```
Comment (rkennke) [inline src/hotspot/os/linux/os_linux.hpp:408]: Minor nit: the e is wrapped onto the next line.

**Review by rkennke — APPROVED** [review]
Looks good to me, just one very minor formatting problem in a comment (doesn't require another review from me).

**Comment by rkennke** [comment]
> x86 test error in compiler/c2/TestVerifyGraphEdges.java unrelated

This should be fixed by #11065.

File: src/hotspot/os/linux/os_linux.hpp [hunk 2]
```diff
@@ -426,6 +392,39 @@ class os::Linux {
...
+  // e malloc'ed more than 4G, so the results from legacy mallinfo() can still be used.
```
Comment (TheRealMDoerr) [inline src/hotspot/os/linux/os_linux.hpp:409]: bad line break

File: test/hotspot/gtest/runtime/test_os_linux.cpp [hunk 3]
```diff
@@ -476,4 +479,29 @@ TEST_VM(os_linux, decoder_get_source_info_invalid) {
...
+  os::Linux::get_mallinfo(&mi, &did_wrap);
+
+  void* p = os::malloc(2 * K, mtTest);
+  ASSERT_NOT_NULL(p);
+
+  // We should see total allocation values > 0
+  ASSERT_GE((mi.uordblks + mi.hblkhd), 2 * K);
+
+  // These values also should not beyond reasonable size, otherwise this indicates bad type
+  // conversion
```
Comment (TheRealMDoerr) [inline test/hotspot/gtest/runtime/test_os_linux.cpp:498]: Incomplete sentence. Is "be" missing?

**Review by TheRealMDoerr — COMMENTED** [review]
LGTM. Can you fix the comments, please?

File: src/hotspot/os/linux/os_linux.hpp (hunk 2 above)
Comment (TheRealMDoerr) [reply src/hotspot/os/linux/os_linux.hpp:409]: Ah, reported by Roman while I was looking at it :-)

File: src/hotspot/os/linux/os_linux.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/os/linux/os_linux.hpp:408]: Huch. Thanks!

**Comment by tstuefe** [comment]
> Hello, besides the little typos already mentioned looks okay.
> But is there really a need to have both structs with same variables/types in it ?

Hi Matthias,

thanks for looking at this.

Both structures have different purposes, though:

- `glibc_mallinfo` in os_linux.hpp holds the output of the JVM internal API `os::Linux::get_mallinfo()`. Its members are large enough to hold the output from both `mallinfo()` and `mallinfo2()`
- `new_mallinfo`, locally in os_linux.cpp, is the output for the glibc function "mallinfo2()". It is defined to make the JVM *buildable* on systems with older glibc, but still *runnable* on a system with a newer glibc. It needs to be binary-compatible with the official glibc "mallinfo2" output structure.

Yes, both are identical, and I considered just using one structure. The old version did this. But I would have to make sure that they are binary compatible and always stay that way too. In the end, I decided to just create separate structures, it seemed less effort, and more obvious to the code reader.

**Review by MBaesken — APPROVED** [review]
LGTM

**Comment by tstuefe** [comment]
Manually tested fastdebug on Alpine and on 32-bit, and merged master. I'll wait for GHAs to complete, then push.

Thanks @dholmes-ora @rkennke @MBaesken for reviewing!

**Comment by tstuefe** [comment]
Manually tested with newer and older glibc, to make sure both code paths work.

**Pull request closed without merging**